dm/syncer/data_validator.go

```diff
@@ -447,119 +610,143 @@ func (v *DataValidator) processRowsEvent(header *replication.EventHeader, ev *re
 		}
 	}
 
 	changeType := getRowChangeType(header.EventType)
 	v.changeEventCount[changeType].Inc()
 
-	columnMap := make(map[string]*model.ColumnInfo)
-	for _, col := range tableInfo.Columns {
-		columnMap[col.Name.O] = col
-	}
-	pk := table.PrimaryKey
-	pkIndices := make([]int, len(pk.Columns))
-	for i, col := range pk.Columns {
-		pkIndices[i] = columnMap[col.Name.O].Offset
-	}
-	table.pkIndices = pkIndices
-
 	step := 1
 	if changeType == rowUpdated {
 		step = 2
 	}
 	for i := 0; i < len(ev.Rows); i += step {
 		row := ev.Rows[i]
-		pkValue := make([]string, len(pk.Columns))
-		for _, idx := range pkIndices {
+		pkValue := make([]string, len(table.pkIndices))
+		for _, idx := range table.pkIndices {
 			pkValue[idx] = genColData(row[idx])
 		}
 		key := genRowKey(pkValue)
 
 		if changeType == rowUpdated {
 			afterRowChangeType := changeType
 			afterRow := ev.Rows[i+1]
-			afterPkValue := make([]string, len(pk.Columns))
-			for _, idx := range pkIndices {
+			afterPkValue := make([]string, len(table.pkIndices))
+			for _, idx := range table.pkIndices {
 				afterPkValue[idx] = genColData(afterRow[idx])
 			}
 			afterKey := genRowKey(afterPkValue)
 			if afterKey != key {
 				// TODO: may reuse IsIdentityUpdated/SplitUpdate of RowChange in pkg/sqlmodel
 				v.dispatchRowChange(key, &rowChange{
-					table:      table,
-					key:        key,
-					pkValues:   pkValue,
-					data:       row,
-					tp:         rowDeleted,
-					lastMeetTS: int64(header.Timestamp),
+					table: table,
+					Key:   key,
+					Data:  row,
+					Tp:    rowDeleted,
 				})
 				afterRowChangeType = rowInsert
 			}
 			v.dispatchRowChange(afterKey, &rowChange{
-				table:      table,
-				key:        afterKey,
-				pkValues:   afterPkValue,
-				data:       afterRow,
-				tp:         afterRowChangeType,
-				lastMeetTS: int64(header.Timestamp),
+				table: table,
+				Key:   afterKey,
+				Data:  afterRow,
+				Tp:    afterRowChangeType,
 			})
 		} else {
 			v.dispatchRowChange(key, &rowChange{
-				table:      table,
-				key:        key,
-				pkValues:   pkValue,
-				data:       row,
-				tp:         changeType,
-				lastMeetTS: int64(header.Timestamp),
+				table: table,
+				Key:   key,
+				Data:  row,
+				Tp:    changeType,
 			})
 		}
 	}
 	return nil
 }
 
+func (v *DataValidator) flushCheckpointAndData(loc binlog.Location) error {
+	var wg sync.WaitGroup
+	wg.Add(v.workerCnt)
+	flushJob := &rowChange{
+		Tp: flushCheckpoint,
+		wg: &wg,
+	}
+	for _, worker := range v.workers {
+		worker.rowChangeCh <- flushJob
+	}
+	wg.Wait()
+
+	err := v.persistHelper.persist(loc)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *DataValidator) loadPersistedData(tctx *tcontext.Context) error {
+	var err error
+	v.location, err = v.persistHelper.loadCheckpoint(tctx)
+	if err != nil {
+		return err
+	}
+
+	var rev int64
+	v.loadedPendingChanges, rev, err = v.persistHelper.loadPendingChange(tctx)
+	if err != nil {
+		return err
+	}
+	// fix missing data
```

Comment:
could you left a comment to explain that which fields should be recovered from loadPendingChange and what to do for else fields?